### PR TITLE
KAFKA-4180 : Authentication with multiple actives Kafka

### PR DESCRIPTION
**Pull request by edoardocomar**
producers/consumers

Changed caching in LoginManager to allow one LoginManager per client
JAAS configuration.
Added test to End2EndAuthorization for SASL Plain and Gssapi with two
consumers with different credentials.

developed with @mimaison

### CODE REVIEW

**Comment by ijuma** [comment]
cc @rajinisivaram

File: core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala]: nit: semi-colon not required

File: core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala]: The bulk of this code is the same as the equivalent test in SaslPlainSslEndToEndAuthorizationTest. Perhaps they could share code?

File: core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala]: The same method could be used for both SaslSslEndToEndAuthorizationTest and SaslPlainSslEndToEndAuthorizationTest to avoid duplication?

File: core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala
Comment (rajinisivaram) [inline core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala]: Should be "no access to group"?

**Review by rajinisivaram — COMMENTED** [review]
Changes look good to me, I have left a few minor comments.

Looks like we are allowing some clients using the static jaas config and other clients using different dynamic jaas config properties to co-exist in the same JVM. Perhaps that should be tested too? Just changing either producer/consumer in one of the tests to use the static config should be sufficient.

**Comment by edoardocomar** [comment]
Hi @ijuma @rajinisivaram could you please review ?
We'd really like this functionality to be included in 10.2 and the freeze is near ;-)

Thanks @rajinisivaram I wrote this and then just saw your comments now

**Comment by edoardocomar** [comment]
Thanks @rajinisivaram I reduced test duplication by introducing an intermediate abstract class in the test hierarchy. I could not avoid having `SslEndToEndAuthorizationTest` as a subclass of `SaslSetup` though due to the connection to zookeeper, but I think this is still cleaner

File: core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala]: moved test to new `abstract class SaslEndToEndAuthorizationTest`

File: core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/SaslSslEndToEndAuthorizationTest.scala]: moved to new `abstract class SaslEndToEndAuthorizationTest`

**Comment by edoardocomar** [comment]
Hi @rajinisivaram about 

> Looks like we are allowing some clients using the static jaas config and other clients using different dynamic jaas config properties to co-exist in the same JVM. Perhaps that should be tested too? Just changing either producer/consumer in one of the tests to use the static config should be sufficient.

I ended up by having the JaasTestSetup statically set the JAAS config for client2
and so the test `testTwoConsumersWithDifferentSASLCredentials` uses the static JAAS config for consumer2 and everything else uses the dynamic config

**Comment by edoardocomar** [comment]
@rajinisivaram @ijuma I'm really keen to have this merged before the feature freeze please could you take a 2nd look, thanks

**Comment by rajinisivaram** [comment]
LGTM

**Comment by edoardocomar** [comment]
@ijuma I had run locally (Mac) the build with JDK 8 and Scala 2.11 and it passes.
Anyway I have rekicked the build by fixing a spurious extra space.

**Comment by ijuma** [comment]
@edoardocomar, you can trigger a new build by simply telling jenkins to retest (we've had that for a few weeks now). I'll take a look at this soon. We should be able to get it into the next release given that it's really a bug fix. Without this, multiple consumers with different jaas configs won't work correctly, right? Given that, we have until the code freeze to merge it.

**Comment by ijuma** [comment]
Btw, one thing that would be good to test: does refreshing of Kerberos tickets work correctly if different credentials are used? We call `kinit -R` in `KerberosLogin` and it would be good to verify how it behaves. Ideally, it will work correctly, but if there are issues, we should document limitations.

**Comment by ijuma** [comment]
@edoardocomar, did you see my question above regarding refreshing of Kerberos tickets?

**Comment by edoardocomar** [comment]
Hi @ijuma we're looking at it, but can't find a test that exercises the ticket refresh. 
We will try a manual test, by setting a low refresh time, and look at the logs to see if they appear OK.

BTW, even if a refresh causes a `reLogin()` to fail, in `KerberosLogin` the method
```
    @Override
    public Subject subject() {
        return subject;
    }
```
is returning the same subject instance obtained by the initial call to `login()`
and not the updated subject from the new `LoginContext` created by `reLogin()`

is this intentional ?

**Comment by ijuma** [comment]
@edoardocomar, yes, we are unfortunately lacking automated tests for that. It's what's blocking this PR of mine: https://github.com/apache/kafka/pull/1522. For this PR, a manual check would be OK.

Regarding the subject, in `reLogin`, we pass the subject to `LoginContext`, so that seems OK.

Finally, seems like there are conflicts in this branch. Would you mind rebasing? I'll try to review it (and possibly merge it) today.

**Comment by edoardocomar** [comment]
Hi @ijuma rebased and included SASL/SCRAM multi-login testing.
Also enhanced the log output of KerberosLogin to be able to distinguish different instances.
We ran the two-consumers test with sleeps and a reduced Kerberos ticket refresh and from the log, it looks OK:

```
[2017-01-10 16:50:04,256] DEBUG Principal=kafka/localhost@EXAMPLE.COM - It is a Kerberos ticket (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:04,256] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT refresh thread started. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:04,257] DEBUG Found TGT with client principal 'kafka/localhost@EXAMPLE.COM' and server principal 'krbtgt/EXAMPLE.COM@EXAMPLE.COM'. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:04,259] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT valid starting at: Tue Jan 10 16:50:04 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:04,259] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT expires: Tue Jan 10 17:00:04 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:04,259] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT refresh sleeping until: Tue Jan 10 16:58:08 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:05,051] DEBUG Principal=client@EXAMPLE.COM - It is a Kerberos ticket (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:05,054] INFO Principal=client@EXAMPLE.COM - TGT refresh thread started. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:05,054] DEBUG Found TGT with client principal 'client@EXAMPLE.COM' and server principal 'krbtgt/EXAMPLE.COM@EXAMPLE.COM'. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:05,054] INFO Principal=client@EXAMPLE.COM - TGT valid starting at: Tue Jan 10 16:50:05 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:05,054] INFO Principal=client@EXAMPLE.COM - TGT expires: Tue Jan 10 17:00:05 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:50:05,055] INFO Principal=client@EXAMPLE.COM - TGT refresh sleeping until: Tue Jan 10 16:58:10 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:08,615] INFO Initiating logout for kafka/localhost@EXAMPLE.COM (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:08,616] INFO Initiating re-login for kafka/localhost@EXAMPLE.COM (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:08,624] DEBUG Found TGT with client principal 'kafka/localhost@EXAMPLE.COM' and server principal 'krbtgt/EXAMPLE.COM@EXAMPLE.COM'. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:08,624] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT valid starting at: Tue Jan 10 16:58:08 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:08,624] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT expires: Tue Jan 10 17:08:08 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:08,624] INFO Principal=kafka/localhost@EXAMPLE.COM - TGT refresh sleeping until: Tue Jan 10 17:06:22 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:10,512] INFO Initiating logout for client@EXAMPLE.COM (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:10,512] INFO Initiating re-login for client@EXAMPLE.COM (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:10,519] DEBUG Found TGT with client principal 'client@EXAMPLE.COM' and server principal 'krbtgt/EXAMPLE.COM@EXAMPLE.COM'. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:10,519] INFO Principal=client@EXAMPLE.COM - TGT valid starting at: Tue Jan 10 16:58:10 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:10,519] INFO Principal=client@EXAMPLE.COM - TGT expires: Tue Jan 10 17:08:10 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 16:58:10,519] INFO Principal=client@EXAMPLE.COM - TGT refresh sleeping until: Tue Jan 10 17:06:29 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 17:01:06,358] DEBUG Principal=client2@EXAMPLE.COM - It is a Kerberos ticket (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 17:01:06,358] INFO Principal=client2@EXAMPLE.COM - TGT refresh thread started. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 17:01:06,359] DEBUG Found TGT with client principal 'client2@EXAMPLE.COM' and server principal 'krbtgt/EXAMPLE.COM@EXAMPLE.COM'. (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 17:01:06,359] INFO Principal=client2@EXAMPLE.COM - TGT valid starting at: Tue Jan 10 17:01:06 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 17:01:06,359] INFO Principal=client2@EXAMPLE.COM - TGT expires: Tue Jan 10 17:11:06 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
[2017-01-10 17:01:06,359] INFO Principal=client2@EXAMPLE.COM - TGT refresh sleeping until: Tue Jan 10 17:09:09 GMT 2017 (org.apache.kafka.common.security.kerberos.KerberosLogin)
```

File: core/src/test/scala/integration/kafka/api/SslEndToEndAuthorizationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/SslEndToEndAuthorizationTest.scala:22]: Some of these imports don't seem to be needed.

File: core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala]: Nit: it should be `Sasl` instead of `SASL` to follow Kafka naming conventions.

File: core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala]: Isn't this timeout a bit high?

File: core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala]: Nit: there should be no space before the `:` in the two lines above.

File: core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/EndToEndAuthorizationTest.scala:336]: Unnecessary new line.

File: clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosLogin.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosLogin.java]: Please use string interpolation. There are a few other places like that.

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java]: Hmm, I think I'd prefer two separate maps with two separate fields in `LoginManager`. It makes things more explicit and easy to understand in my opinion (even though it's a bit more code).

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java]: This is an existing bug, but `mode` should be `loginType`.

**Review by ijuma — CHANGES_REQUESTED** [review]
Thanks for doing the test @edoardocomar. I've left some comments. I think they should be easy to address.

File: core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala
Comment (edoardocomar) [reply core/src/test/scala/integration/kafka/api/SaslEndToEndAuthorizationTest.scala]: Yes! the Junit time that Eclipse displayed for the single test method is actually inclusive of setup/teardown. The actual test method is not so long.

File: clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosLogin.java
Comment (edoardocomar) [reply clients/src/main/java/org/apache/kafka/common/security/kerberos/KerberosLogin.java]: done

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java
Comment (edoardocomar) [reply clients/src/main/java/org/apache/kafka/common/security/authenticator/LoginManager.java]: Honestly, I am not sure the code with two maps is better ...

**Comment by edoardocomar** [comment]
Thanks @ijuma - Honestly, I am not sure the LoginManager with two maps is better ...

**Comment by edoardocomar** [comment]
Please retest this

**Comment by edoardocomar** [comment]
retest this please

**Review by ijuma — APPROVED** [review]
Thanks for the PR, LGTM. I made a few tweaks before merging to trunk:
1. Principal prefix is more consistent with how we do it in other parts of the code.
2. Simplified `acquireLoginManager` a bit.
3. Removed unused `clientKeyTabFile` var.
4. Increased test timeouts (on second thought, it's better for them not to be too low because of Jenkins)
5. Renamed some test files for better consistency with other files.

**Pull request closed without merging**
